File: java/src/main/java/abstract_factory/product/SuperBus.java
package abstract_factory.product;

import abstract_factory.product.part.VehiclePartFactory;

public class SuperBus extends Vehicle {
    public static final String NAME = "Super Bus";
    private final VehiclePartFactory vehiclePartFactory;

    public SuperBus(VehiclePartFactory vehiclePartFactory) {
        name = NAME;
        this.vehiclePartFactory = vehiclePartFactory;
    }

    @Override
    public void assembly() {
        engine = vehiclePartFactory.createEngine();
        tire = vehiclePartFactory.createTire();
        options = vehiclePartFactory.createOptions();
    }
}

File: src/main/kotlin/abstract_factory/product/SuperBus.kt
package abstract_factory.product

import abstract_factory.product.part.VehiclePartFactory

class SuperBus(private val vehiclePartFactory: VehiclePartFactory) : Vehicle() {
    override var name = NAME

    override fun assembly() {
        engine = vehiclePartFactory.createEngine()
        tire = vehiclePartFactory.createTire()
        options = vehiclePartFactory.createOptions()
    }

    companion object {
        const val NAME: String = "Super Bus"
    }
}